pkg/controller/common/grafana_api_impl.go

```diff
@@ -0,0 +1,183 @@
+		return resp, err
+	}
+	fmt.Printf("response == %s\n", *datasourceResp.Message)
+
+	resp.Succeeded = true
+	datasource.Status.ID = *(datasourceResp.ID)
+	return resp, nil
+}
+func (h GrafanaApiData) GetDataSource(ctx context.Context, datasource *v1alpha1.GrafanaDataSource) (ResponseStatus, error) {
+	resp := ResponseStatus{}
+	client, err := h.getClient(datasource.Spec.GrafanaRef)
+	if err != nil {
+		return resp, err
+	}
+	_, err = client.GetDatasourceByName(datasource.Spec.DataSource.Name)
+	if err != nil {
+		return resp, err
+	}
+	return resp, errors.New("not implemented")
+}
+func (h GrafanaApiData) DeleteDataSource(ctx context.Context, datasource *v1alpha1.GrafanaDataSource) (ResponseStatus, error) {
+	resp := ResponseStatus{}
+	client, err := h.getClient(datasource.Spec.GrafanaRef)
+	if err != nil {
+		return resp, err
+	}
+	_, err = client.DeleteDatasource(datasource.Status.ID)
+	if err != nil {
+		fmt.Printf("Failed to set dashboard: %+v", err)
+		return resp, err
+	}
+
+	resp.Succeeded = true
+	return resp, nil
+}
+
+func (h GrafanaApiData) DataSourceIsKnown(ctx context.Context, datasource *v1alpha1.GrafanaDataSource) (bool, error) {
+	// custom resource holds multiple datasources,
+	// DataSourceIsKnown return true only if all datasources are present in api
+	client, err := h.getClient(datasource.Spec.GrafanaRef)
+	if err != nil {
+		return false, err
+	}
+	_, err = client.GetDatasourceByName(datasource.Spec.DataSource.Name)
+	if err != nil {
+		fmt.Printf("Datasource is not Knonwn, err = %v\n", err)
+		return false, nil
+	}
+	fmt.Printf("Datasource is Knonwn")
+	return true, nil
+}
+
+func dataSourceFromApi(from *v1alpha1.GrafanaDataSource) sdk.Datasource {
+	return sdk.Datasource{
+		OrgID:             1,
+		Name:              from.Spec.DataSource.Name,
+		Type:              from.Spec.DataSource.Type,
+		Access:            from.Spec.DataSource.Access,
+		URL:               from.Spec.DataSource.Url,
+		Password:          &from.Spec.DataSource.Password,
+		User:              &from.Spec.DataSource.User,
+		Database:          &from.Spec.DataSource.Database,
+		BasicAuth:         &from.Spec.DataSource.BasicAuth,
+		BasicAuthUser:     &from.Spec.DataSource.BasicAuthUser,
+		BasicAuthPassword: &from.Spec.DataSource.BasicAuthPassword,
+		IsDefault:         from.Spec.DataSource.IsDefault,
+		JSONData:          from.Spec.DataSource.JsonData,
+	}
+}
+
+func (h GrafanaApiData) getClient(grafanaName string) (*sdk.Client, error) {
+	// currently supports grafana in same namespace as operator only
+
+	if val, ok := h.clients[grafanaName]; ok {
+		return val, nil
+	}
+	grafana, err := h.kubeHelper.grafanaClient.Grafanas(os.Getenv("WATCH_NAMESPACE")).Get(grafanaName, v1.GetOptions{})
+	if err != nil {
+		fmt.Printf("Failed to get grafana %#v", err)
+		return nil, err
+	}
+	url := "http://" + GrafanaServiceName + "." + grafana.Namespace + ":3000/"
+	creds := grafana.Spec.AdminUser + ":" + grafana.Spec.AdminPassword
```

Comment:
Currently the `adminUser` and `adminPassword` fields in the CR are optional and Grafana uses default credentials if not set. In that case this would not work. I'd like to make this more reliable by removing the options entirely from the CR and have the operator set up an admin user by default. The credentials of the admin user should then be written to the CR status.

Reply:
I was actually thinking about it and what I'd like to see eventually is something like this:
1. create grafana (potentially on 127.0.0.1 only)
2. create api token using basic auth (`kubectl exec curl -XPOST -u user:password 127.0.0.1/api/auth/keys`?)
3. save api token as secret
4. redeploy grafana with basic auth disabled
5. expose grafana (create service and ingress resource)

What I can do right now is to hardcode default credentials (admin:admin) as default values.

What do you think about that?